redis/src/commands/cluster_scan.rs

```diff
@@ -0,0 +1,709 @@
+                    scan_state = new_scan_state;
+                    (new_cursor, new_keys) = from_redis_value(&scan_result).unwrap();
+                }
+                Ok((Err(err), _)) => return Err(err),
+                Err(err) => return Err(err),
+            }
+        }
+    };
+
+    // If the cursor is 0, meaning we finished scanning the address
+    // we will update the scan state to get the next address to scan
+    if new_cursor == 0 {
+        scan_state = match scan_state
+            .update_scan_state_and_get_next_address(&core)
+            .await
+        {
+            Ok(state) => state,
+
+            Err(err) => return Err(err),
+        }
+    };
+
+    // If the address is empty, meaning we finished scanning all the addresss
+    if scan_state.address_in_scan.is_empty() {
+        return Ok((ScanStateCursor::new(), new_keys));
+    }
+
+    scan_state = ScanState::create(
+        new_cursor,
+        scan_state.scanned_slots_map,
+        scan_state.address_in_scan,
+        scan_state.address_epoch,
+    );
+    Ok((ScanStateCursor::from_scan_state(&scan_state), new_keys))
+}
+
+// Send the scan command to the address in the scan_state
+async fn send_scan<C>(
+    scan_state: &ScanState,
+    core: &C,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+) -> RedisResult<Value>
+where
+    C: ClusterInScan,
+{
+    let mut scan_command = cmd("SCAN");
+    scan_command.arg(scan_state.cursor);
+    if let Some(match_pattern) = match_pattern {
+        scan_command.arg("MATCH").arg(match_pattern);
+    }
+    if let Some(count) = count {
+        scan_command.arg("COUNT").arg(count);
+    }
+    if let Some(object_type) = object_type {
+        scan_command.arg("TYPE").arg(object_type.to_string());
+    }
+    core.route_command(&scan_command, &scan_state.address_in_scan)
+        .await
+}
+
+// If the scan command faild to route to the address we will check we will first refresh the slots, we will check if all slts are coverd by cluster,
+// and if so we will try to get a new address to scan for handeling case of failover.
+// if all slots are not coverd by the cluster we will return an error indicating that the cluster is not well configured.
+// if all slots are coverd by cluster but we failed to get a new address to scan we will return an error indicating that we failed to get a new address to scan.
+// if we got a new address to scan but the scan command failed to route to the address we will return an error indicating that we failed to route the command.
+async fn retry_scan<C>(
+    scan_state: &ScanState,
+    core: &C,
+    match_pattern: Option<String>,
+    count: Option<usize>,
+    object_type: Option<ObjectType>,
+) -> RedisResult<(RedisResult<Value>, ScanState)>
+where
+    C: ClusterInScan,
+{
+    let refresh_result = core.refresh_slots().await;
+    match refresh_result {
+        Ok(_) => {
+            if !core.is_all_slots_covered().await {
```

Comment:
so you can't scan over a partially covered cluster? IMO that's the wrong behavior. Start scanning, the cluster might become complete as you scan.
And if it remains partial throughout the scan, just return the results. This is in line with the behavior of the cluster connection in other commands that are sent to all the nodes.

Reply:
So that's true - but we decided that its add a lot of complexity to the command since we based on slots so we need to check for coverage and somehow mark missing slots as missing and check whether that's changed for each node scan.
We think that's clusters with missing slots are rare - for clusters to keep functioning with missing slots the user needs implicitly change the config of the cluster and this will make just minor commands to work - and the cluster can't be fixed if the shards that failed recovered. Meaning you can't recover by adding a new node since the cluster will not agree to add a new node in such a state.
So the scenario is cluster without replicas that one of the node crushed but will recover at some point, and all of this during scan.
that pretty rare and we decided that the complication of the code by it is not worth it.